#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/RectangleArena.proto"

WorldInfo {
}
Viewpoint {
  orientation 0.31063873544537396 -0.6009155471314888 -0.7364808763685331 5.059164408709177
  position -0.6977207103299642 -1.417718578384647 0.9433249571512239
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
  floorSize 2 2
  floorAppearance PBRAppearance {
    roughness 1
    roughnessMap ImageTexture {
      url [
        "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/textures/parquetry/mosaic_parquetry_roughness.jpg"
      ]
    }
    metalness 0
    normalMap ImageTexture {
      url [
        "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/textures/parquetry/mosaic_parquetry_normal.jpg"
      ]
    }
    occlusionMap ImageTexture {
      url [
        "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/appearances/protos/textures/parquetry/mosaic_parquetry_occlusion.jpg"
      ]
    }
  }
}
Solid {
  translation 0.45 -0.73 0.05
  rotation 0 0 -1 -0.13090030610042527
  children [
    DEF obstacle1 Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  boundingObject USE obstacle1
  locked TRUE
}
Solid {
  translation 0.666748 -0.198009 0.05
  rotation 0 0 -1 -1.3089953071795861
  children [
    DEF obstacle2 Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.2 0.1 0.1
      }
    }
  ]
  name "solid(1)"
  boundingObject USE obstacle2
  locked TRUE
}
Solid {
  translation 0.36358 0.381911 0.05
  rotation 0 0 -1 -1.047195307179586
  children [
    DEF obstacle3 Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.12 0.1 0.1
      }
    }
  ]
  name "solid(3)"
  boundingObject USE obstacle3
  locked TRUE
}
Solid {
  translation -0.380764 0.661551 0.05
  rotation 0 0 1 2.09439
  children [
    DEF obstacle4 Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.12 0.1 0.1
      }
    }
  ]
  name "solid(4)"
  boundingObject USE obstacle4
  locked TRUE
}
Solid {
  translation -0.545793 -0.0756207 0.05
  rotation 0 0 1 1.83259
  children [
    DEF obstacle5 Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.1 0.3 0.1
      }
    }
  ]
  name "solid(5)"
  boundingObject USE obstacle5
  locked TRUE
}
Solid {
  translation -0.356546 -0.527723 0
  rotation 0 0 1 2.8797896938995744
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 1 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.03 0.51 0.0005
      }
    }
  ]
  name "solid(6)"
  locked TRUE
}
Solid {
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Mesh {
        url [
          "../path/outer_path.obj"
        ]
      }
    }
  ]
  name "solid(2)"
  boundingObject Mesh {
    url [
      "../path/outer_path.obj"
    ]
  }
  locked TRUE
}
Solid {
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Mesh {
        url [
          "../path/inner_path.obj"
        ]
      }
    }
  ]
  name "solid(7)"
  boundingObject Mesh {
    url [
      "../path/inner_path.obj"
    ]
  }
  locked TRUE
}
Robot {
  translation -0.11005441617887361 -0.5142097224458014 0.03828626005424963
  rotation -0.5293499065416939 0.5913940266224588 0.6083105964222341 -2.144383911991415
  children [
    Gyro {
      translation 0.02 0.02 0.03
      children [
        Shape {
          appearance PBRAppearance {
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
    }
    Accelerometer {
      translation 0 0.04 -0.05
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0 1 0.7
          }
          geometry Box {
            size 0.05 0.05 0.05
          }
        }
      ]
    }
    InertialUnit {
      translation -0.03 0.02 0
      children [
        Shape {
          appearance PBRAppearance {
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "inertial_unit"
    }
    Lidar {
      translation 2.67257e-05 0.0299963 0.000272275
      rotation 0.5030470370970558 0.6020562948954822 0.6200579781315597 -2.19
      name "lidar(1)"
      type "rotating"
    }
    Pose {
      children [
        DEF body Shape {
          appearance PBRAppearance {
            baseColor 0.5 0.6 1
          }
          geometry Box {
            size 0.1 0.04 0.16
          }
        }
      ]
    }
    Pose {
      translation 0 0 0.08
      children [
        Lidar {
          translation 0 0 0.08
          rotation 0 1 0 0
          horizontalResolution 1
          fieldOfView 6.28319
          numberOfLayers 1
          maxRange 5
          type "rotating"
          noise 0.01
          resolution 512
        }
        Shape {
          appearance PBRAppearance {
            baseColor 0.8 0.8 0.8
          }
          geometry Cylinder {
            height 0.02
            radius 0.02
          }
        }
      ]
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        position 33.405387255875134
        anchor 0.0598 -0.017 -0.05
      }
      device [
        RotationalMotor {
          name "left_wheel_motor"
          maxTorque 0.1
        }
        PositionSensor {
          name "left_wheel_sensor"
          resolution 0.01
        }
      ]
      endPoint Solid {
        translation 0.059800001312102835 -0.01698331557968466 -0.04999952704361235
        rotation 0.6425687610497809 -0.41738665123157687 -0.64256810588189 2.350783047075523
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColor 1 0 0
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.02
              radius 0.02
            }
          }
        ]
        boundingObject USE wheel
        physics Physics {
          mass 0.05
        }
        linearVelocity -3.763975974026405e-09 7.206955323886938e-10 9.248310988141308e-09
        angularVelocity -1.9806653310791177e-08 -1.783460822407756e-07 -8.237772117243798e-09
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        position 21.342179661275907
        anchor -0.0598 -0.017 -0.05
      }
      device [
        RotationalMotor {
          name "right_wheel_motor"
          maxTorque 0.1
        }
        PositionSensor {
          name "right_wheel_sensor"
          resolution 0.01
        }
      ]
      endPoint Solid {
        translation -0.05979999861182306 -0.016983315974136704 -0.04999952274966629
        rotation 0.6879158968047286 -0.23139663537838007 -0.6879151953975947 2.686803309650409
        children [
          USE wheel
        ]
        boundingObject USE wheel
        physics Physics {
          mass 0.05
        }
        linearVelocity -2.7881309684101195e-09 6.123156528320803e-10 9.248210659506776e-09
        angularVelocity -1.9806675985764082e-08 -1.7834606550409956e-07 -8.237771291272846e-09
      }
    }
    BallJoint {
      jointParameters BallJointParameters {
        position -37.284928457914525
        anchor 0 -0.03 0.05
      }
      endPoint Solid {
        translation 8.242889255605235e-07 -0.029946912142796548 0.050002208210074624
        rotation -0.5798257970531574 -0.51892459870249 0.6281077184155044 -1.2991831927338362
        children [
          DEF caster Shape {
            appearance PBRAppearance {
              baseColor 0 0 0
              roughness 1
              metalness 0
            }
            geometry Sphere {
              radius 0.01
            }
          }
        ]
        boundingObject USE caster
        physics Physics {
        }
        linearVelocity -1.3654981902306195e-09 1.3001023176779194e-09 -7.083731386544657e-12
        angularVelocity -2.381570239978109e-07 -2.018377492585384e-08 7.2096343806903e-11
      }
    }
  ]
  boundingObject USE body
  physics Physics {
    mass 0.3
  }
  controller "wall_following_controller"
  linearVelocity -6.508997400437809e-09 1.436577507984637e-09 -8.862327357899162e-09
  angularVelocity -1.9806656422343574e-08 -1.783460783449949e-07 -8.237743231481612e-09
}
